### PR TITLE
Support mixed PBC neighbor list 

### PR DESCRIPTION
**Pull request by curtischong**
- This PR implements the cpu and cuda logic to support mixed PBC
- I added a few tests in python and c++ to help me verify correctness
- I did not make the fortran tests pass since I don't know fortran and am not comfortable modifying it

### CODE REVIEW

File: vesin/tests/neighbors.cpp [hunk 1]
```diff
@@ -131,6 +132,50 @@ TEST_CASE("Non-periodic") {
     );
 }
 
+TEST_CASE("Mixed periodic boundaries") {
```
Comment (curtischong) [inline vesin/tests/neighbors.cpp]: this test isn't technically required since we have the python tests

File: torch/src/vesin_torch.cpp [hunk 2]
```diff
@@ -381,7 +393,16 @@ std::vector<torch::Tensor> AutogradNeighbors::backward(
...
+    auto periodic_cpu = periodic;
+    if (!periodic_cpu.device().is_cpu()) {
+        periodic_cpu = periodic_cpu.to(torch::kCPU);
+    }
+    if (!periodic_cpu.is_contiguous()) {
+        periodic_cpu = periodic_cpu.contiguous();
+    }
+    const bool* periodic_data = periodic_cpu.data_ptr<bool>();
+    bool any_periodic = periodic_data[0] || periodic_data[1] || periodic_data[2];
+    if (any_periodic && box.requires_grad()) {
```
Comment (curtischong) [inline torch/src/vesin_torch.cpp]: I think this is fine because we've already calculated the new vectors_grad based on periodicity. so we don't need to perform this logic on a per-periodic axis manner.

(if two points are not neighbors because of nonperiodicity, their vectors_grad will just be 0).

Not 100% sure though.

File: python/vesin/vesin/_neighbors.py [hunk 3]
```diff
@@ -37,7 +37,7 @@ def compute(
         self,
         points: "npt.ArrayLike",
         box: "npt.ArrayLike",
-        periodic: bool,
+        periodic: "Union[bool, npt.ArrayLike]",
```
Comment (curtischong) [inline python/vesin/vesin/_neighbors.py]: I decided to make this param a union of bool and npt arraylike to prevent this from being a breaking change for existing users. Note that this doesn't allow list[bool]. I'm not opinionated about not having list[bool] but decided to put up the PR with this minimal change first.

File: vesin/src/cpu_cell_list.cpp [hunk 4]
```diff
@@ -234,19 +233,34 @@ void CellList::foreach_pair(Function callback) {
...
+            if (is_neighboring_cell_outside_bounds) {
+                continue;
+            }
 
             for (const auto& atom_i: current_cell) {
                 for (const auto& atom_j: this->get_cell(neighbor_cell_i)) {
                     auto shift = CellShift{cell_shift} + atom_i.shift - atom_j.shift;
                     auto shift_is_zero = shift[0] == 0 && shift[1] == 0 && shift[2] == 0;
 
-                    if (!box_.periodic() && !shift_is_zero) {
```
[the remark this answers is not in the document]
Comment (Luthaf) [reply vesin/src/cpu_cell_list.cpp]: Do you have any benchmarks regarding the speed difference, or is this based on gut feeling from the code changes? I went with a different approach, but I could also port this as an optimization.
